### ISSUE TITLE
Better packing of data with best-fit decrease strategy

### ISSUE THREAD

**Issue by ntnq4** [huggingface/trl#4554]
Hello,

When using packing with the bfd strategy, it looks like too much truncation is done when the seq_length is smaller than the average length of the sequences we want to pack.

For example : 
```python
from datasets import Dataset
from trl import pack_dataset

examples = {
    "input_ids": [[1, 2, 3, 4], [5, 6], [7, 8, 9], [10]],
    "attention_mask": [[1, 1, 1, 1], [1, 0], [1, 0, 0], [1]],
}
dataset = Dataset.from_dict(examples)

packed_dataset = pack_dataset(dataset, seq_length=3, strategy="bfd")
print(packed_dataset )
```

results in:
```python
{'input_ids': [[1, 2, 3], [7, 8, 9], [5, 6, 10]],
 'attention_mask': [[1, 1, 1], [1, 0, 0], [1, 0, 1]],
 'seq_lengths': [[3], [3], [2, 1]]}
```
So the token '4' is missing from the training tokens.

In a extreme case:
```python
examples_2 = {
    "input_ids": [[0, 0], [1, 2, 3, 4], [5, 6, 7, 8, 9], [10]],
    "attention_mask": [[1, 1], [1, 1, 1, 1], [1, 1, 1, 1, 1], [1]],
}
dataset_2 = Dataset.from_dict(examples_2)
print(pack_dataset(dataset_2, seq_length=1, strategy="bfd")[:])
```
results in:
```python
{'input_ids': [[0], [1], [5], [10]],
 'attention_mask': [[1], [1], [1], [1]],
 'seq_lengths': [[1], [1], [1], [1]]}
```

So here we are basically applying truncation to every sequence instead of having twelve sequences of one token.

If we put ourself in a more usefull setting, when I was finetunning on some very long sequences with a seq_lenfth of 4096, the majority of the tokens was discarded y the bfd packing. On my dataset, the bfd method kept only 0.2% of the total training tokens.

Is the behavior normal ?
I would find it useful to add an option to still have tokens that are deleted in other sequences, even if this is less than ideal. It would be a good compromise between the current versions of bfd and wrapped.

**Comment by qgallouedec** [comment]
Thanks, very well explained.
Yes it's expected. We were hesitant between these two designs. Are you able to compare these two setup? (I'm not even sure if it's easy to fairly compare)

**Comment by ntnq4** [comment]
Thanks for the quick feedback.
For now I only tested bfd versus wrapped.

Here is a table for a quick comparaison on a fine tuning I made on a huge dataset for 5 epochs with the same hyperparameterrs except the packing method:

|   | bfd | wrapped |
|---|---|---|
|  time |  30.15min | 1.192 day  | 
| train loss  |  0.615 |  0.409 |  
|  test loss  | 0.604  |  0.451 |  
|  train/num_tokens  | 26k  |  1B |  

As I'm training on code data that are formated in a repo-level way, I have very long sequences that do not fit with my compute budget so I was using a 4096 seq_length.  I'm aware that this setup can be very different from the common use of trl. 

We can see the gap between those two versions but we absolutely cannot compare it as the number of token seen if very different.

I didn't experience a setup of bfd where I could keep the residues that were removed for now but it would be interested to compare it to the wrapped version to see the performances with the same amount of tokens.

**Comment by jiosephlee** [comment]
Perhaps it's worth raising a warning when this happens? It seems like the reasonable thing would be to let the user know this is an issue, so they can manually chunk, etc. accordingly
